File: frontend/src/api/units.ts
import { apiClient } from './client'
import type { Unit, UnitCreate, UnitKind } from '../types/unit'

export const unitsApi = {
  list: (params?: { kind?: UnitKind; active?: boolean; skip?: number; limit?: number }) => {
    const q = new URLSearchParams()
    if (params?.kind) q.set('kind', params.kind)
    if (params?.active !== undefined) q.set('active', String(params.active))
    if (params?.skip !== undefined) q.set('skip', String(params.skip))
    if (params?.limit !== undefined) q.set('limit', String(params.limit))
    const qs = q.toString()
    return apiClient.get<Unit[]>(`/units/${qs ? `?${qs}` : ''}`)
  },
  get: (id: number) => apiClient.get<Unit>(`/units/${id}`),
  create: (unit: UnitCreate) => apiClient.post<Unit>('/units/', unit),
  update: (id: number, unit: UnitCreate) => apiClient.put<Unit>(`/units/${id}`, unit),
  remove: (id: number) => apiClient.delete<void>(`/units/${id}`),
}

File: backend/app/api/routes/units.py
from typing import Optional

from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, require_roles
from app.core.database import get_db
from app.core.roles import ADMIN_ROLES
from app.schemas.unit import UnitCreate, UnitKind, UnitOut
from app.services import unit_service

router = APIRouter(
    prefix="/units",
    tags=["units"],
    dependencies=[Depends(get_current_user)],
)


@router.post(
    "/",
    response_model=UnitOut,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_roles(*ADMIN_ROLES))],
)
def create_unit(unit_in: UnitCreate, db: Session = Depends(get_db)):
    return unit_service.create_unit(db, unit_in)


@router.get("/", response_model=list[UnitOut], status_code=status.HTTP_200_OK)
def list_units(
    skip: int = 0,
    limit: int = 200,
    kind: Optional[UnitKind] = None,
    active: Optional[bool] = None,
    db: Session = Depends(get_db),
):
    return unit_service.list_units(db, skip=skip, limit=limit, kind=kind, active=active)


@router.get("/{unit_id}", response_model=UnitOut, status_code=status.HTTP_200_OK)
def get_unit(unit_id: int, db: Session = Depends(get_db)):
    return unit_service.get_unit(db, unit_id)


@router.put(
    "/{unit_id}",
    response_model=UnitOut,
    status_code=status.HTTP_200_OK,
    dependencies=[Depends(require_roles(*ADMIN_ROLES))],
)
def update_unit(unit_id: int, unit_in: UnitCreate, db: Session = Depends(get_db)):
    return unit_service.update_unit(db, unit_id, unit_in)


@router.delete(
    "/{unit_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(require_roles(*ADMIN_ROLES))],
)
def delete_unit(unit_id: int, db: Session = Depends(get_db)):
    unit_service.delete_unit(db, unit_id)
    return None
